sound/soc/sof/trace.c

```diff
@@ -13,11 +13,173 @@
 #include "sof-priv.h"
 #include "ops.h"
 
+#define TRACE_FILTER_ELEMENT_EXPANSION 10
+#define TRACE_FILTER_CHARS_PER_ELEMENT_GUESS 4
+
+static int trace_filter_append_elem(int32_t key, int32_t value,
+				    struct sof_ipc_trace_filter_elem **elem_list,
+				    size_t *capacity, size_t *counter)
+{
+	struct sof_ipc_trace_filter_elem *elem_ptr;
+	size_t new_capacity;
+
+	/* assert enough space, allocate space for a few more elements to reduce krealloc() calls */
```

Comment:
@ktrzcinx I think we can optimize this a bit by not having any need for krealloc at all. The user input is known before-hand with each element separated by a ;. If we use strsep() , it would tell you exactly how many elems you'd allocate space for.

Reply:
and then you could use a nested strsep() with the space as the delimiter to keep searching within each element. This would eliminate the need for the per element guess as well

Reply:
>each element separated by a ;

I can easily count logger entries, because they are separated by a `;`, each entry contains at least 1, up to 4 IPC elements, depending on content (trace_filter_parse_entry() shows it exactly), so I need to fully parse input to know exactly how many IPC elements are inside. Even so, I can assume worst case (4 IPC elements per each logger entry) and then remove part responsible for capacity extension.